File: platform/polyglot/rag_llm_synthesis/mjs/src/reranker.mjs
/**
 * @fileoverview Gemini-based listwise reranker.
 */

const GEMINI_BASE_URL = 'https://generativelanguage.googleapis.com/v1beta/openai/';

/**
 * Rerank candidates using Gemini as a listwise reranker.
 *
 * Each candidate is a [document, score] tuple where document has a
 * pageContent (or page_content) property.
 *
 * @param {string} query - The search query
 * @param {Array<[{pageContent?: string, page_content?: string}, number]>} candidates
 * @param {number} k - Number of top results to return
 * @param {Object} [opts]
 * @param {string} [opts.apiKey] - Gemini API key (falls back to GEMINI_API_KEY env)
 * @param {string} [opts.model='gemini-2.0-flash'] - Gemini model name
 * @returns {Promise<Array<[any, number]>>}
 */
export async function geminiRerank(query, candidates, k, opts = {}) {
  if (!candidates.length) {
    return candidates;
  }

  const cap = Math.min(candidates.length, 30);
  const capped = candidates.slice(0, cap);

  let chunksText = '';
  for (let i = 0; i < capped.length; i++) {
    const [doc] = capped[i];
    const content = (doc.pageContent || doc.page_content || '').slice(0, 600).replace(/\n/g, ' ').trim();
    chunksText += `[${i}] ${content}\n\n`;
  }

  const prompt =
    `You are a relevance ranker. Given a query and ${cap} text chunks, ` +
    `rank them by relevance to the query. Return ONLY a JSON array of indices ` +
    `from most relevant to least relevant. Example: [3, 0, 7, 1, ...]\n\n` +
    `Query: ${query}\n\nChunks:\n${chunksText}`;

  try {
    const apiKey = opts.apiKey || process.env.GEMINI_API_KEY;
    if (!apiKey) {
      return candidates.slice(0, k);
    }

    const model = opts.model || 'gemini-2.0-flash';
    const resp = await fetch(`${GEMINI_BASE_URL}chat/completions`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model,
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.0,
      }),
    });

    if (!resp.ok) {
      return candidates.slice(0, k);
    }

    const data = await resp.json();
    const content = data.choices[0].message.content.trim();

    let indices;
    try {
      indices = JSON.parse(content);
    } catch {
      const match = content.match(/\[[\d,\s]+\]/);
      if (match) {
        indices = JSON.parse(match[0]);
      } else {
        return candidates.slice(0, k);
      }
    }

    if (!Array.isArray(indices)) {
      return candidates.slice(0, k);
    }

    const reranked = [];
    const seen = new Set();
    for (const idx of indices) {
      if (Number.isInteger(idx) && idx >= 0 && idx < cap && !seen.has(idx)) {
        seen.add(idx);
        reranked.push(capped[idx]);
      }
    }
    for (let i = 0; i < capped.length; i++) {
      if (!seen.has(i)) {
        reranked.push(capped[i]);
      }
    }

    return reranked.slice(0, k);
  } catch {
    return candidates.slice(0, k);
  }
}

File: platform/polyglot/rag_llm_synthesis/py/rag_llm_synthesis/reranker.py
"""Gemini-based listwise reranker."""

import json
import os
import re
from typing import List, Tuple, Any, Optional


def gemini_rerank(
    query: str,
    candidates: List[Tuple[Any, float]],
    k: int,
    api_key: Optional[str] = None,
    model: str = "gemini-2.0-flash",
) -> List[Tuple[Any, float]]:
    """Rerank candidates using Gemini as a listwise reranker.

    Each candidate is a (document, score) tuple where document has a
    .page_content attribute.

    Args:
        query: The search query.
        candidates: List of (document, score) tuples to rerank.
        k: Number of top results to return.
        api_key: Gemini API key. Falls back to GEMINI_API_KEY env var.
        model: Gemini model name for reranking.

    Returns:
        Reranked list of (document, score) tuples, limited to k results.
    """
    if not candidates:
        return candidates

    cap = min(len(candidates), 30)
    capped = candidates[:cap]

    chunks_text = ""
    for i, (doc, _score) in enumerate(capped):
        snippet = doc.page_content[:600].replace("\n", " ").strip()
        chunks_text += f"[{i}] {snippet}\n\n"

    prompt = (
        f"You are a relevance ranker. Given a query and {cap} text chunks, "
        f"rank them by relevance to the query. Return ONLY a JSON array of indices "
        f"from most relevant to least relevant. Example: [3, 0, 7, 1, ...]\n\n"
        f"Query: {query}\n\nChunks:\n{chunks_text}"
    )

    try:
        from openai import OpenAI

        _api_key = api_key or os.environ.get("GEMINI_API_KEY")
        if not _api_key:
            return candidates[:k]

        client = OpenAI(
            api_key=_api_key,
            base_url="https://generativelanguage.googleapis.com/v1beta/openai/",
        )
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.0,
        )
        content = response.choices[0].message.content.strip()

        try:
            indices = json.loads(content)
        except json.JSONDecodeError:
            match = re.search(r'\[[\d,\s]+\]', content)
            if match:
                indices = json.loads(match.group())
            else:
                return candidates[:k]

        if not isinstance(indices, list):
            return candidates[:k]

        reranked = []
        seen = set()
        for idx in indices:
            if isinstance(idx, int) and 0 <= idx < cap and idx not in seen:
                seen.add(idx)
                reranked.append(capped[idx])
        for i, c in enumerate(capped):
            if i not in seen:
                reranked.append(c)

        return reranked[:k]

    except Exception:
        return candidates[:k]
